src/site/content/en/blog/kapwing/index.md

```diff
@@ -0,0 +1,440 @@
+  const trak = demuxer.moov.traks.find((trak: any) => trak.tkhd.track_id === trak_id);
+  const data = await demuxer.getFrameDataRange(
+    preceedingKeyFrameIndex,
+    desiredSampleIndex
+  );  
+  /* end demuxer interface */
+
+  for (let i = preceedingKeyFrameIndex; i <= desiredSampleIndex; i += 1) {
+    const sample = trak.samples[i];
+    const sampleData = data.readNBytes(
+      sample.offset,
+      sample.size
+    );  
+
+    const sampleType = sample.is_sync ? 'key' : 'delta';
+    const encodedFrame = new EncodedVideoChunk({
+      sampleType,
+      timestamp: sample.cts,
+      duration: sample.duration,
+      samapleData,
+    }); 
+
+    if (i === desiredSampleIndex)
+      timestampToMatch = encodedFrame.timestamp;
+    decoder.decodeEncodedFrame(encodedFrame, i); 
+  }
+  await decoder.flush();
+
+  return { type: 'value', value: decodedSample };
+};
+```
+
+The structure of the demuxer is quite complex and beyond the scope of this 
+article. It stores each frame in an array titled `samples`. We use the demuxer 
+to find the closest preceding key frame to our desired timestamp, which is 
+where we must begin video decode.
+
+Videos are composed of full frames known as key or i-frames as well as much 
+smaller delta frames, often referred to as p- or b-frames. Decode must always 
+begin at a key frame. 
+
+The application decodes frames by:
+
+1. Instantiating the decoder with a frame output callback.
+1. Configuring the decoder for the specific codec and input resolution.
+1. Creating an `encodedVideoChunk` using data from the demuxer
+1. Calling the `decodeEncodedFrame` method
+
+We do this until we reach the frame with the desired timestamp.
+
+## What's next?
+
+We define scale on our frontend as the ability to maintain precise and 
+performant playback as projects get larger and more complex. One way to scale 
+performance is to mount as few videos as possible at once, however when we do 
+this we risk slow and choppy transitions. While we have developed internal 
+systems to cache video components for reuse, there are limitations to how much 
+control HTML5 video tags can provide.
+
+In the future, we may attempt to play all media using WebCodecs. This could 
+allow us to be very precise about what data we buffer which should help scale 
+performance.
+
+We can also do a better job of offloading large trackpad computations to web 
+workers, and can be more intelligent about pre-fetching files and 
+pre-generating frames. We see large opportunities to optimize our overall 
+application performance, and to extend functionality with tools like webGL.
+
+We would also like to continue our investment in tensorflow.js, which we 
+currently use for intelligent background removal, and which we plan to 
+leverage for other sophisticated tasks such as object detection, feature 
+extraction, style transfer and so on.
+
+We're excited about the promise of native like performance and functionality 
```

Comment:
@heyawhite  Can you try something to smooth out this abrupt change of tone?

If you have time for it, it's not absolutely necessary, it's just a nice finishing touch and an opportunity for a writing exercise.